control-plane/agents/src/bin/core/volume/specs.rs

```diff
@@ -748,19 +748,33 @@ impl ResourceSpecsLocked {
     pub(crate) fn get_or_create_volume(
         &self,
         request: &CreateVolumeSource,
-    ) -> ResourceMutex<VolumeSpec> {
+    ) -> Result<ResourceMutex<VolumeSpec>, SvcError> {
         let mut specs = self.write();
         if let Some(volume) = specs.volumes.get(&request.source().uuid) {
-            volume.clone()
+            Ok(volume.clone())
         } else {
+            // if request has a capacity limit, add up the volumes and reject
+            // if the capacity limit would be exceeded
+            match request.source().capacity_limit {
+                None => {} // no limit, no check needed
+                Some(limit) => {
+                    let mut total: u64 = request.source().size;
+                    for vol in specs.volumes.to_vec() {
+                        total += vol.lock().size;
+                        if total > limit {
+                            return Err(SvcError::CapacityLimitExceeded {});
+                        }
+                    }
```

Comment:
Don't need to clone the list, can just take ref, and sum the sizes:
```suggestion
                    let total: u64 = specs.volumes.values().map(|v| v.lock().size).sum();
                    if total > limit {
                        return Err(SvcError::CapacityLimitExceeded {});
                    }
```

Reply:
fixed